internal/antewrapper/msg_fees_decorator.go

```diff
@@ -0,0 +1,240 @@
+	}
+
+	if !requiredFees.IsZero() && !feeCoins.IsAnyGTE(requiredFees) {
+		return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, "Base Fee+additional fee cannot be paid with fee value passed in "+": %q, required: %q = %q(base-fee) +%q(additional-fees)", feeCoinsOriginal, requiredFees.Add(additionalFees...),
+			requiredFees, additionalFees)
+	}
+
+	return nil
+}
+
+func EnsureAccountHasSufficientFeesWithAcctBalanceCheck(gas uint64, feeCoins sdk.Coins, additionalFees sdk.Coins,
+	balancePerCoin sdk.Coins, minGasPriceForAdditionalFeeCalc sdk.Coin) error {
+	err := EnsureSufficientFees(gas, feeCoins, additionalFees, minGasPriceForAdditionalFeeCalc)
+	if err != nil {
+		return err
+	}
+
+	if balancePerCoin.IsZero() || !balancePerCoin.IsAnyGTE(feeCoins) {
+		return fmt.Errorf("fee payer account does not have enough balance to pay for %q", feeCoins)
+	}
+	return nil
+}
+
+// EnsureSufficientFees to be used by msg_service_router
+func EnsureSufficientFees(gas uint64, feeCoins sdk.Coins, additionalFees sdk.Coins,
+	minGasPriceForAdditionalFeeCalc sdk.Coin) error {
+	// Step 1. Check if fees has enough money to pay additional fees.
+	var hasNeg bool
+	if feeCoins, hasNeg = feeCoins.SafeSub(additionalFees); hasNeg {
+		return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, DefaultInsufficientFeeMsg+": %q, required additional fee: %q", feeCoins, additionalFees)
+	}
+	// Step 2: check if additional fees in nhash, that base fees and additional fees can be paid
+	// total fees in hash - gas limit * price per gas >= additional fees in hash
+	if containsDenom(additionalFees, minGasPriceForAdditionalFeeCalc.Denom) {
+		// Determine the required fees by multiplying each required minimum gas
+		// price by the gas limit, where fee = ceil(minGasPrice * gasLimit).
+		fee := minGasPriceForAdditionalFeeCalc.Amount.Mul(sdk.NewIntFromUint64(gas))
+		baseFees := sdk.NewCoin(minGasPriceForAdditionalFeeCalc.Denom, fee)
+		if feeCoins, hasNeg = feeCoins.SafeSub(sdk.Coins{baseFees}); hasNeg {
+			return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, DefaultInsufficientFeeMsg+": %q, required additional fee: %q", feeCoins, additionalFees)
+		}
+	}
+
+	return nil
+}
+
+// CalculateAdditionalFeesToBePaid computes the stability tax on MsgSend and MsgMultiSend.
+func CalculateAdditionalFeesToBePaid(ctx sdk.Context, mbfk msgfeestypes.MsgFeesKeeper, msgs ...sdk.Msg) (sdk.Coins, error) {
+	// get the msg fee
+	additionalFees := sdk.Coins{}
+
+	for _, msg := range msgs {
+		typeURL := sdk.MsgTypeURL(msg)
+		msgFees, err := mbfk.GetMsgFee(ctx, typeURL)
+		if err != nil {
+			return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, err.Error())
+		}
+
+		if msgFees == nil {
+			continue
+		}
+		if msgFees.AdditionalFee.IsPositive() {
+			additionalFees = additionalFees.Add(sdk.NewCoin(msgFees.AdditionalFee.Denom, msgFees.AdditionalFee.Amount))
```

Comment:
If AdditionalFee is already a coin, why are we creating a new coin around it here?

Reply:
good catch ..changed